crates/loro-internal/src/loro.rs

```diff
@@ -243,14 +276,18 @@ impl LoroDoc {
         Option<CommitOptions>,
         Option<LoroMutexGuard<Option<Transaction>>>,
     ) {
-        let mut txn_guard = self.txn.lock().unwrap();
-        if !self.auto_commit.load(Acquire) {
-            // if not auto_commit, nothing should happen
-            // because the global txn is not used
-            return (None, Some(txn_guard));
+        {
+            let txn_guard = self.txn.lock().unwrap();
+            if !self.auto_commit.load(Acquire) {
+                // if not auto_commit, nothing should happen
+                // because the global txn is not used
+                return (None, Some(txn_guard));
+            }
         }
 
         loop {
+            self.before_commit();
```

Comment:
We need to call this inside `txn.commit` method

Reply:
But it's kinda tricky... Because we would need to release the txn lock before calling before_commit. But if it's inside `txn.commit`, we need to hold the lock of txn...